src/toolkits/object_detection/object_detector.cpp

```diff
@@ -379,35 +379,86 @@ void object_detector::train(gl_sframe data,
-// existing backend is available?
-variant_map_type object_detector::perform_evaluation(gl_sframe data,
-                                                     std::string metric) {
   std::vector<std::string> metrics;
   static constexpr char AP[] = "average_precision";
   static constexpr char MAP[] = "mean_average_precision";
   static constexpr char AP50[] = "average_precision_50";
   static constexpr char MAP50[] = "mean_average_precision_50";
-  std::vector<std::string> all_metrics = {AP,MAP,AP50,MAP50};
-  if (std::find(all_metrics.begin(), all_metrics.end(), metric) != all_metrics.end()) {
+  std::vector<std::string> all_metrics = {AP, MAP, AP50, MAP50};
+  if (std::find(all_metrics.begin(), all_metrics.end(), metric) !=
+      all_metrics.end()) {
     metrics = {metric};
+  } else if (metric == "auto") {
+    metrics = {AP50, MAP50};
+  } else if (metric == "all" || metric == "report") {
+    metrics = all_metrics;
+  } else {
+    log_and_throw("Metric " + metric + " not supported");
   }
-  else if (metric == "auto") {
-    metrics = {AP50,MAP50};
+
+  flex_list class_labels = read_state<flex_list>("classes");
+
+  // Initialize the metric calculator
+  average_precision_calculator calculator(class_labels);
+
+  auto consumer = [&](const std::vector<image_annotation>& predicted_row,
+                      const std::vector<image_annotation>& groundtruth_row) {
+    calculator.add_row(predicted_row, groundtruth_row);
+  };
+
+  perform_predict(data, consumer);
+
+  // Compute the average precision (area under the precision-recall curve) for
+  // each combination of IOU threshold and class label.
+  variant_map_type result_map = calculator.evaluate();
+
+  // Trim undesired metrics from the final result. (For consistency with other
+  // toolkits. In this case, almost all of the work is shared across metrics.)
+  if (std::find(metrics.begin(), metrics.end(), AP) == metrics.end()) {
+    result_map.erase(AP);
   }
-  else if (metric == "all" || metric == "report") {
-    metrics = all_metrics;
+  if (std::find(metrics.begin(), metrics.end(), AP50) == metrics.end()) {
+    result_map.erase(AP50);
   }
-  else {
-    log_and_throw("Metric " + metric + " not supported");
+  if (std::find(metrics.begin(), metrics.end(), MAP50) == metrics.end()) {
+    result_map.erase(MAP50);
+  }
+  if (std::find(metrics.begin(), metrics.end(), MAP) == metrics.end()) {
+    result_map.erase(MAP);
   }
 
+  return result_map;
+}
+
+gl_sarray object_detector::predict(gl_sframe data) {
+  gl_sarray_writer result(flex_type_enum::LIST, 1);
+
+  auto consumer = [&](const std::vector<image_annotation>& predicted_row,
+                      const std::vector<image_annotation>& groundtruth_row) {
+    // Convert predicted_row to flex_type list to call gl_sarray_writer
+    flex_list predicted_row_ft;
+    for (auto const& each_row : predicted_row) {
+      flex_list fl = {each_row.bounding_box.x, each_row.bounding_box.y,
```

Comment:
The bounding box should be a dictionary, not a list. Reference the parsing code from https://github.com/apple/turicreate/blob/becc15cb1bd32cbcae50d0d2fbb65810508b51c9/src/toolkits/object_detection/od_data_iterator.cpp#L53